test/functional/shada/registers_spec.lua

```diff
@@ -160,4 +160,16 @@ describe('ShaDa support code', function()
     reset()
     eq({{'\171'}, 'v'}, getreg('e'))
   end)
+
+  it('remembers which register was the unnamed register when loading',
+  function()
+    setreg('0', {'zero'}, 'c')
+    setreg('1', {'one'}, 'cu')
+    setreg('2', {'two'}, 'c')
+    nvim_command('qall')
+    reset()
+    eq({{'zero'}, 'v'}, getreg('0'))
+    eq({{'one'}, 'v'}, getreg('1'))
+    eq({{'one'}, 'v'}, getreg('"'))
+  end)
```

Comment:
This matches the behavior in Vim as far as I can tell. Doing:

```VimL
call setreg('0', 'This is in reg 0', 'c')
echo getreg('"')
call setreg('3', 'This is in reg 3', 'c')
echo getreg('"')
```

Results in `This is in reg 0` being printed twice.

While I haven't tested the hypothetical case where the viminfo/shada file has something in register `0` but no info on the unnamed register, that won't occur and make a difference unless someone manually edits the file. I have the feeling that the behavior should be the same.